### PR TITLE
[BEAM-22] Add ShardControlledWrite override

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace "<Jira issue #>" in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This is used for TextIO and AvroIO, which provide withNumOutputShards
methods to control the number of output files. Apply this override in
the InProcessPipelineRunner.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ */
+abstract class ShardControlledWrite<InputT>
+    extends ForwardingPTransform<PCollection<InputT>, PDone> {
+  @Override
+  public PDone apply(PCollection<InputT> input) {
+    int numShards = getNumShards();
+    checkArgument(
+        numShards >= 1,
+        "%s should only be applied if the output has a controlled number of shards (> 1); got %s",
+        ShardControlledWrite.class.getSimpleName(),
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java]: swap to this.getClass() to get the actual implementation name vs ShardControlledWrite.class

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  abstract PTransform<? super PCollection<InputT>, PDone> getSingleShardTransform(int shardNum);
+
+  private static class RandomPartitionFn<T> implements Partition.PartitionFn<T> {
+    private Random random;
+    @Override
+    public int partitionFor(T elem, int numPartitions) {
+      if (random == null) {
+        // lazily initialize at execution time
+        random = new SecureRandom();
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java]: Use a regular random, or checkout ThreadLocalRandom
I don't think we need cryptographically good randomness

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/AvroIOShardedWriteFactory.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/*
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/AvroIOShardedWriteFactory.java]: Please add inprocess runner specific tests to validate sharding transform application/execution

**Comment by lukecwik** [comment]
Minor comments, other then need tests

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ShardControlledWrite.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/AvroIOShardedWriteFactory.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/AvroIOShardedWriteFactory.java]: I've added tests for the expansion.

**Comment by tgroh** [comment]
I've also changed the partitionFn, to evenly distribute a bundle across partitions and randomly select a "first partition".

**Comment by lukecwik** [comment]
Jenkins & Travis failed

**Comment by tgroh** [comment]
Fixed.

**Comment by lukecwik** [comment]
LGTM

**Pull request closed without merging**